zerver/lib/markdown/__init__.py

```diff
@@ -2181,10 +2181,10 @@ def build_inlinepatterns(self) -> markdown.util.Registry:
         # Custom bold syntax: **foo** but not __foo__
         # str inside ** must start and end with a word character
         # it need for things like "const char *x = (char *)y"
-        EMPHASIS_RE = r"(\*)(?!\s+)([^\*^\n]+)(?<!\s)\*"
+        EMPHASIS_RE = r"(\*)(?!\s+)([^\*^\n]+?)(?<!\s)\2"
         STRONG_RE = r"(\*\*)([^\n]+?)\2"
-        STRONG_EM_RE = r"(\*\*\*)(?!\s+)([^\*^\n]+)(?<!\s)\*\*\*"
-        TEX_RE = r"\B(?<!\$)\$\$(?P<body>[^\n_$](\\\$|[^$\n])*)\$\$(?!\$)\B"
+        STRONG_EM_RE = r"(\*{3})(?!\s+)([^\*^\n]+?)(?<!\s)\2"
```

Comment:
Can you explain why you added the `?`s?  That would seem to suggest that a block that only contains whitespace should be captured?

Reply:
@timabbott, I've added `?` as:
- this is more common in upstream: [example1](https://github.com/Python-Markdown/markdown/blob/faf73b08bb984669fcba044259f8d67fbfed69ab/markdown/inlinepatterns.py#L142), [example2](https://github.com/Python-Markdown/markdown/blob/faf73b08bb984669fcba044259f8d67fbfed69ab/markdown/inlinepatterns.py#L118).
- `+?` is a non-greedy qualifier so having regex: [`(\*{3})(?!\s+)([^\*^\n]+)(?<!\s)\2`](https://regex101.com/r/cItDwp/1) will be equivalent to [`(\*{3})(?!\s+)([^^\n]+?)(?<!\s)\2`](https://regex101.com/r/N8ZGpL/1).
- We follow the same pattern in [`DEL_RE`](https://github.com/zulip/zulip/blob/1d4b6c13207086a212cdfc30af26bdbe9a600a4b/zerver/lib/markdown/__init__.py#L2180).

> That would seem to suggest that a block that only contains whitespace should be captured?

No, please have a look at the examples: [example1](https://regex101.com/r/GvWmRs/1), [example2](https://regex101.com/r/lgLsFT/1).

Reply:
> * this is more common in upstream: [example1](https://github.com/Python-Markdown/markdown/blob/faf73b08bb984669fcba044259f8d67fbfed69ab/markdown/inlinepatterns.py#L142), [example2](https://github.com/Python-Markdown/markdown/blob/faf73b08bb984669fcba044259f8d67fbfed69ab/markdown/inlinepatterns.py#L118).

It’s needed in those upstream examples because they’re `.+?` rather than a more carefully selected character set. The ones you’ve added here make no difference.

> * `+?` is a non-greedy qualifier so having regex: [`(\*{3})(?!\s+)([^\*^\n]+)(?<!\s)\2`](https://regex101.com/r/cItDwp/1) will be equivalent to [`(\*{3})(?!\s+)([^^\n]+?)(?<!\s)\2`](https://regex101.com/r/N8ZGpL/1).

That is not how a non-greedy qualifier works, and those regexes are not equivalent. Try running them on `***strong**strong***` to see the difference.

Reply:
> It’s needed in those upstream examples because they’re `.+?` rather than a more carefully selected character set. The ones you’ve added here make no difference.

Ok, in such a case, we should consider modifying [`DEL_RE = r"(?<!~)(\~\~)([^~\n]+?)(\~\~)(?!~)"`](https://github.com/zulip/zulip/blob/afe64041677a93dda6d203122885f7599931d467/zerver/lib/markdown/__init__.py#L2192) to `DEL_RE = r"(?<!~)(\~\~)([^~\n]+)(\~\~)(?!~)"`?

> those regexes are not equivalent. Try running them on `***strong**strong***` to see the difference.

Oops, actually I forgot to exclude `\*` in the latter one (`[^^\n]`); excluding `\*` should work.
The example: `***strong**strong***` is still not matched by the current [`STRONG_EM_RE = r"(\*\*\*)(?!\s+)([^\*^\n]+)(?<!\s)\*\*\*"`](https://github.com/zulip/zulip/blob/afe64041677a93dda6d203122885f7599931d467/zerver/lib/markdown/__init__.py#L2198)